### PR TITLE
[Merged by Bors] - Optimize finalized chain sync by skipping newPayload messages

### PR DESCRIPTION
**Pull request by GeemoCandama**
## Issue Addressed

#3704 

## Proposed Changes
Adds is_syncing_finalized: bool parameter for block verification functions. Sets the payload_verification_status to Optimistic if is_syncing_finalized is true. Uses SyncState in NetworkGlobals in BeaconProcessor to retrieve the syncing status.

## Additional Info
I could implement FinalizedSignatureVerifiedBlock if you think it would be nicer.

### CODE REVIEW

File: beacon_node/http_api/src/lib.rs
Comment (michaelsproul) [inline beacon_node/http_api/src/lib.rs]: This will be evaluated once when the API server is built and passed to every API function from that point on, which isn't quite what we want (the sync state could change).

I think actually for the API server we can always set `is_syncing_finalized=false`, because the blocks from the API are new blocks which should be at the head of the chain (they're not from sync at all, let alone finalized sync).

**Review by michaelsproul — COMMENTED** [review]
This is awesome, thanks for taking it on!

We have a convention of not using `bool` parameters for arguments, particular where they affect important parts of block verification. The reasoning is that it's too easy to accidentally mix up multiple bool params, or to mix up `true` and `false`.

It might be a pain to change, but it would be great if we could use a dedicated `enum` like:

```rust
/// Signal whether the execution payloads of new blocks should be
/// immediately verified with the EL or imported optimistically without
/// any EL communication.
pub enum NotifyExecutionLayer {
   #[default]
   Yes,
   No,
}
```

Hopefully your editor/IDE/terminal can help you with the rename+refactor.

The enum also gives us the flexibility of setting `NotifyExecutionLayer::No` for reasons other than the blocks being part of finalization sync, e.g. in future we could have a flag like `--optimistic-sync=always` that allows one to run an optimistically synced Lighthouse node without any execution client connected (although that would be very spicy and not I'm not recommending that for this PR :hot_pepper:).

File: beacon_node/http_api/src/lib.rs
Comment (GeemoCandama) [reply beacon_node/http_api/src/lib.rs]: Thank you for reviewing this! Before using the boolean I tried passing the SyncState from BeaconProcessor as the parameter, but I ran into some weird lifetime issues I couldn't quite figure out. Do you think that would be a good approach or should I stick with what you suggested above?

Comment (GeemoCandama) [reply beacon_node/http_api/src/lib.rs]: Nevermind your comment kind of answers that for me I'll implement what you suggested. There is already the AllowOptimisticImport enum I should use that right?

Comment (michaelsproul) [reply beacon_node/http_api/src/lib.rs]: I think `AllowOptimisticImport` serves a slightly different purpose, I'd prefer we add a new enum just for this use, the `NotifyExecutionLayer` one. In the API we can hardcode `NotifyExecutionLayer::Yes`

File: beacon_node/http_api/src/lib.rs
Comment (michaelsproul) [inline beacon_node/http_api/src/lib.rs]: we may as well leave this file as-is:

```
git checkout origin/unstable -- beacon_node/http_api/src/lib.rs
```

**Comment by GeemoCandama** [comment]
I think this is ready. I know you all are busy with the new version, but I just wanted to mark this as finished from my perspective.

File: beacon_node/network/src/beacon_processor/mod.rs
Comment (michaelsproul) [inline beacon_node/network/src/beacon_processor/mod.rs]: This is a minor point, but I think we might _only_ want to compute this for `ChainSegment`. That's more conservative (we'll always verify RPC and Gossip blocks) while still achieving our goal of speeding up sync. It also saves us locking the `sync_state` for every type of message processed, e.g. attestations, where we don't actually need it.

**Review by michaelsproul — COMMENTED** [review]
Looks good.

I'd like to do some more manual testing before we merge.

We're also blocked on v3.3.0 release currently

File: beacon_node/network/src/beacon_processor/mod.rs
Comment (GeemoCandama) [reply beacon_node/network/src/beacon_processor/mod.rs]: Ah, that does seem better! I will make the changes.

Comment (GeemoCandama) [reply beacon_node/network/src/beacon_processor/mod.rs]: I was playing around with making the new type FinalizedSignatureVerifiedBlock like you had suggested, since then we could remove the notify_execution_layer parameter from a lot of the functions. Because all the fields of SignatureVerifiedBlock are private I was having trouble making the SignatureVerifiedBlock into a Finalized one. I could have added getter methods to SignatureVerifiedBlock, but I assumed that would be undesirable. Do you think I should make the new type since there is really only one path I saw in the code where notify_execution_layer: NotifyExecutionLayer is actually dependent on the sync_state?

Comment (GeemoCandama) [reply beacon_node/network/src/beacon_processor/mod.rs]: Why do the CI tests seem to fail non-deterministically? In the other PR I'm working on this same group of tests failed: test-suite / release-tests-windows (pull_request). Is there something I can do to ensure things pass?

**Review by michaelsproul — APPROVED** [review]
This is great, thanks.

I did some manual testing and found very healthy sync speeds running this PR, peaking at 16 slots/second rather than ~5-6 slots/second for `unstable`. Looking at the metrics shows a complete absence of `newPayload` messages during finalized chain sync and an average of 1 fcU every 5 seconds.

**Comment by paulhauner** [comment]
It's been said here before, but this is great! Thank you @GeemoCandama 🙏

**Comment by michaelsproul** [comment]
Adding this to the merge queue!

bors r+

**Pull request closed without merging**
